Add Minio component

tkislan opened the pull request:
## Description:
Reference issue: https://github.com/home-assistant/home-assistant/issues/20187

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#9267

## Example entry for `configuration.yaml` (if applicable):
```yaml
minio:
  host: localhost
  port: 9000
  access_key: SO9KNO6YT9OGE39PQCZW
  secret_key: EzH5iGs3qbJr1MwIMf2AUyfqzewgd2qgrk9+i+vU
  secure: false
  listen:
    - bucket: test
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] [_The manifest file_][manifest-docs] has all fields filled out correctly ([example][ex-manifest]).
  - [x] New dependencies have been added to `requirements` in the manifest ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-manifest]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json
[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/mobile_app/manifest.json#L5
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23
[manifest-docs]: https://developers.home-assistant.io/docs/en/development_checklist.html#_the-manifest-file_

tkislan committed:
Add minio implementation

tkislan committed:
Static check changes

tkislan committed:
Added docstrings

tkislan committed:
Update docstrings

tkislan committed:
Update docstrings

tkislan committed:
Fix linter errors

tkislan committed:
Finally fix all docstring errors

tkislan committed:
Create services.yaml

tkislan committed:
Update CODEOWNERS

tkislan committed:
Final changes

tkislan committed:
Remove double underscores

tkislan committed:
Merge remote-tracking branch 'origin/dev' into minio

tkislan committed:
Minor changes

tkislan committed:
Update config.yml

tkislan commented on homeassistant/components/minio/minio_helper.py:
```diff
@@ -0,0 +1,208 @@
+    }
+    # noinspection PyProtectedMember
+    # pylint: disable=W0212
+    return minio_client._url_open(
```
@MartinHjelmare 
Convo from previous PR.
https://github.com/minio/minio-py/blob/7107c84183cf5fb4deff68c0a16ab9f1c0b4c37e/minio/api.py#L491
Implementation in minio library wraps everything, and only yields retrieved events, and gives no access to the underlying request.
So there is no way to interrupt this, and cleanly stop the listener.

MartinHjelmare replied:
It would be better to make a PR to the minio library with a change so it's possible to use the library cleanly.

tkislan replied:
Indeed I agree. I will create PR there as well, and will update it here once it gets merged into upstream.

MartinHjelmare commented on homeassistant/components/minio/__init__.py:
```diff
@@ -0,0 +1,280 @@
+    )
+
+    def _render_service_value(service, key):
+        value = service.data.get(key)
```
Use `dict[key]`.

MartinHjelmare commented on homeassistant/components/minio/minio_helper.py:
```diff
@@ -0,0 +1,208 @@
+"""Minio helper methods."""
+
```
Please remove this blank line.

tkislan committed:
Review changes

tkislan committed:
Added tests

MartinHjelmare commented on tests/components/minio/test_minio.py:
```diff
@@ -0,0 +1,53 @@
+import homeassistant.components.minio as minio
+
+
+class TestMinioHelper(unittest.TestCase):
```
All new tests should be standalone pytest test functions. See the smartthings test suite for example.

tkislan committed:
Fix lint errors

tkislan committed:
Move tests from unittest to pytest

tkislan committed:
Add minio as test requirement

tkislan committed:
Update test_minio_helper.py

tkislan committed:
Merge remote-tracking branch 'upstream/dev' into dev

tkislan committed:
Merge remote-tracking branch 'origin/dev' into minio

tkislan commented:
Resolved all the comments, and also updated against latest upstream dev.

MartinHjelmare commented on homeassistant/components/minio/__init__.py:
```diff
@@ -0,0 +1,282 @@
+                event['bucket'],
+                event['key']
+            )
+            self._hass.bus.async_fire(DOMAIN, {
```
Don't use the async api. That can only be used inside the event loop thread. Here we're in a separate thread.

tkislan replied:
But for some reason the test fails when using sync `.fire()` method. As the `await hass.async_block_till_done()` doesn't wait for the listen callback to be called

MartinHjelmare replied:
We have to solve that in the test.

tkislan commented on tests/components/minio/test_minio.py:
```diff
@@ -0,0 +1,44 @@
+
+    @ha.callback
+    def listener(event):
+        print("FUCK")
```
Ops. Forgot to remove this. Was trying to track down why my callback is being fired after the test already finished

MartinHjelmare replied:
Remove the debug print.

tkislan committed:
Better event thread handling, added hass test

MartinHjelmare commented on tests/components/minio/test_minio.py:
```diff
@@ -1,44 +1,195 @@
+            }
+        )
+
+        hass.bus.fire(EVENT_HOMEASSISTANT_START)
```
Use the async api in async context. We're in a coroutine here, ie async context.

We can await `hass.async_start` instead.

MartinHjelmare commented on tests/components/minio/test_minio.py:
```diff
@@ -0,0 +1,195 @@
+
+    minio_client_mock = MagicMock()
+
+    with patch('homeassistant.components.minio.Minio') as minio_mock:
```
I would break this out into a pytest fixture function in which we yield the client mock.

MartinHjelmare commented on tests/components/minio/test_minio.py:
```diff
@@ -0,0 +1,195 @@
+        },
+        blocking=True
+    )
+    minio_client_mock.fput_object.assert_called_with(
```
Use:
```py
from unittest.mock import call
...
assert minio_client_mock.fput_object.call_args == call(...)
```
Since the mock object isn't specced, a typo in the assert method would be interpretred as an attribute of the mock.

MartinHjelmare commented on tests/components/minio/test_minio.py:
```diff
@@ -0,0 +1,195 @@
+        while not events:
+            await asyncio.sleep(0)
+
+        hass.bus.fire(EVENT_HOMEASSISTANT_STOP)
```
Await `hass.async_stop`.

tkislan committed:
Update tests

tkislan committed:
Fixed lint errors

tkislan committed:
Update test_minio.py

tkislan commented:
@MartinHjelmare
Is anything blocking this?

MartinHjelmare commented on homeassistant/components/minio/minio_helper.py:
```diff
@@ -0,0 +1,235 @@
+                )
+            except json.JSONDecodeError:
+                response.close()
+            except AttributeError:
```
Why would the `AttributeError` happen?

tkislan replied:
Added comment

MartinHjelmare commented on homeassistant/components/minio/minio_helper.py:
```diff
@@ -0,0 +1,235 @@
+                break
+
+            # Wait before attempting to connect again.
+            time.sleep(1)
```
Is there a timeout when trying to connect? What happens if the resource is unavailable? Do we need a back off reconnect plan?

tkislan replied:
This is not handling resource unavailability. Just the case, when endpoint closes the longpolling response any reason.
Failing to connect will actually resolve to closing the thread, and not reconnect anymore. Will actually address this, and try to reconnect as well.

robmarkcole replied:
That sounds reasonable to me

dgomes replied:
This is actually throttling the long polling response...

tkislan replied:
Moved it into the relevant except block

MartinHjelmare commented on homeassistant/components/minio/minio_helper.py:
```diff
@@ -0,0 +1,235 @@
+                    )
+                # Fail gracefully. If for whatever reason this stops working,
+                # it shouldn't prevent it from firing events.
+                # pylint: disable=W0703
```
Please disable the same line only if possible without crossing the 79 character per line, and use words to describe the error to disable: `broad-except`.

MartinHjelmare commented on tests/components/minio/test_minio.py:
```diff
@@ -0,0 +1,202 @@
+    """Patch helper function for minio notification stream."""
+    with patch(
+        'homeassistant.components.minio'
+        '.minio_helper.get_minio_notification_response'
```
We shouldn't patch our function, but we should patch the minio client function that `get_minio_notification_response` uses.

MartinHjelmare commented on tests/components/minio/test_minio.py:
```diff
@@ -0,0 +1,202 @@
+    assert 0 == len(event.data['metadata'])
+
+
+async def test_queue_listener():
```
Do we need to test this specifically? Isn't this already tested in the test above?

tkislan replied:
It's not the same test. This is a unit test, testing a particular functionality.
The other one is more of an integration test. So in a way, it is also tested there, but it is in no way the same test.

MartinHjelmare replied:
It's the other test that is important. We don't really care that the implementation should be in a specific way. I think we can remove these tests. They just decrease robustness.

tkislan replied:
It actually helps you to pin point the issue faster. Please enlighten me how more (or more granular) tests decrease robustness ...

MartinHjelmare replied:
We can't change the implementation without changing the test. That makes the test brittle.

tkislan replied:
What? That the actual point of the test. You never want to have implementation in the tests, as that just adds another place, where bug can be introduced ...

robmarkcole replied:
I can see both sides of the argument, but I think what is important in the long run is maintainability. If the potential downsides of needing to maintain the test outweigh the current benefit of having the test, then we can omit it

tkislan replied:
Removed the tests ...

MartinHjelmare commented on tests/components/minio/test_minio_helper.py:
```diff
@@ -0,0 +1,40 @@
+from tests.components.minio.common import TEST_EVENT
+
+
+async def test_event_object_iteration(hass):
```
See above.

MartinHjelmare commented on tests/components/minio/test_minio_helper.py:
```diff
@@ -0,0 +1,40 @@
+    assert 0 == len(metadata)
+
+
+async def test_minio_event_stream_iterator(hass):
```
See above.

tkislan committed:
Review changes

MartinHjelmare commented on homeassistant/components/minio/__init__.py:
```diff
@@ -60,12 +63,12 @@
 BUCKET_KEY_SCHEMA = vol.Schema({
-    vol.Required('bucket'): cv.template,
-    vol.Required('key'): cv.template,
+    vol.Required(CONF_BUCKET): cv.template,
```
We only call config keys `CONF_`. Service data keys we call `ATTR_`.

MartinHjelmare commented on tests/components/minio/test_minio_helper.py:
```diff
@@ -0,0 +1,40 @@
+
+    response_mock.stream.return_value = stream_mock
+
+    stream_mock.__next__.side_effect = [
```
I think we should incorporate different replies in the first test instead.

tkislan committed:
More review changes

robmarkcole commented:
Chaps what happened to this PR - looks very close to merge?

MartinHjelmare commented:
I'm happy to hear a second opinion from another member.

robmarkcole commented:
OK I am sitting on the fence with my comment about the tests but if thats all thats holding this up lets just make a decision and go with it

dgomes commented on homeassistant/components/minio/minio_helper.py:
```diff
@@ -0,0 +1,242 @@
+
+        while True:
+            if self._should_stop:
+                break
```
`while not self._should_stop:` ?

dgomes commented on homeassistant/components/minio/minio_helper.py:
```diff
@@ -0,0 +1,242 @@
+            self._event_stream_it.close()
+            self._event_stream_it = None
+
+        _LOGGER.info('Joining event thread')
```
debug instead of info?

dgomes commented on homeassistant/components/minio/minio_helper.py:
```diff
@@ -0,0 +1,242 @@
+        )
+
+        if not bucket or not key:
+            _LOGGER.info('Invalid bucket and/or key, %s, %s', bucket, key)
```
warn ?

dgomes reviewed (commented):
I'm with @MartinHjelmare on the test issue

robmarkcole commented:
ping @tkislan just a couple of changes required

tkislan committed:
Merge remote-tracking branch 'upstream/dev' into dev

tkislan committed:
Merge branch 'dev' into minio

tkislan committed:
Removed tests

tkislan committed:
Applied code style changes

tkislan committed:
Reformat test code

tkislan commented:
Removed the tests in question ...
Updated formatting

Tests are currently failing because `bimmer-connected` package version 0.5.6 got taken down from pypi for some reason

tkislan committed:
Merge remote-tracking branch 'upstream/dev' into dev

tkislan committed:
Merge branch 'dev' into minio

MartinHjelmare reviewed (approved):
Thanks!
